File: 1072-flip-columns-for-maximum-number-of-equal-rows/1072-flip-columns-for-maximum-number-of-equal-rows.py
from collections import Counter

class Solution:
    def maxEqualRowsAfterFlips(self, matrix):
        freq = Counter()
        for row in matrix:
            first_val = row[0]
            pattern = ''.join('A' if val == first_val else 'B' for val in row)
            freq[pattern] += 1
        return max(freq.values())

File: 1072-flip-columns-for-maximum-number-of-equal-rows/1072-flip-columns-for-maximum-number-of-equal-rows.java
class Solution {
    public int maxEqualRowsAfterFlips(int[][] matrix) {
        Map<String, Integer> freq = new HashMap<>();
        int maxlen = 0;
        for (int[] row : matrix) {
            StringBuilder pattern = new StringBuilder();
            int firstVal = row[0];
            for (int val : row) {
                pattern.append(val == firstVal ? "A" : "B");
            }
            String key = pattern.toString();
            freq.put(key, freq.getOrDefault(key, 0) + 1);
        }
        for (int count : freq.values()) {
            maxlen = Math.max(maxlen, count);
        }
        return maxlen;
    }
}
